istioctl/pkg/validate/validate.go

```diff
@@ -0,0 +1,186 @@
+	if exists {
+		obj, err := crd.ConvertObjectFromUnstructured(schema, un, "")
+		if err != nil {
+			return fmt.Errorf("cannot parse proto message: %v", err)
+		}
+		return schema.Validate(obj.Name, obj.Namespace, obj.Spec)
+	} else {
+		ev := &store.BackendEvent{
+			Key: store.Key{
+				Name:      un.GetName(),
+				Namespace: un.GetNamespace(),
+				Kind:      un.GetKind(),
+			},
+			Value: mixerCrd.ToBackEndResource(un),
+		}
+		return mixerValidator.Validate(ev)
+	}
+}
+
+type NoGroupError struct{ msg string }
+type NoVersionError struct{ msg string }
+type NoKindError struct{ msg string }
+type NoNameError struct{ msg string }
+type DecodeError struct{ msg string }
+type ValidationError struct{ msg string }
+
+func (e NoGroupError) Error() string    { return e.msg }
+func (e NoVersionError) Error() string  { return e.msg }
+func (e NoKindError) Error() string     { return e.msg }
+func (e NoNameError) Error() string     { return e.msg }
+func (e DecodeError) Error() string     { return e.msg }
+func (e ValidationError) Error() string { return e.msg }
+
+func validateObjectStream(reader io.Reader, writer io.Writer) error {
+	// We store configs as a YaML stream; there may be more than one decoder.
+	yamlDecoder := kubeyaml.NewYAMLOrJSONDecoder(reader, 512*1024)
+	for i := 0; ; i++ {
+		// We want to skip empty documents in the stream of yaml documents.
+		// unstructured decoding expects valid metadata, e.g. kind. This
+		// results in false positive decode errors when we encounter empty
+		// docs. Decode into the underlying JSON map directly first and create
+		// a valid unstructured.Unstructured with non-empty docs.
+		var content map[string]interface{}
+		if err := yamlDecoder.Decode(&content); err != nil {
+			if err == io.EOF {
+				return nil // validation complete
+			}
+			return &DecodeError{fmt.Sprintf("cannot decode yaml: %v", err)}
+		}
+		if len(content) == 0 {
+			continue
+		}
+
+		un := &unstructured.Unstructured{}
+		un.SetUnstructuredContent(content)
+
+		gvk := un.GetObjectKind().GroupVersionKind()
+		if len(gvk.Group) == 0 {
+			return &NoGroupError{fmt.Sprintf("resource #%v has no group", i)}
+		} else if len(gvk.Version) == 0 {
+			return &NoVersionError{fmt.Sprintf("resource #%v has no version", i)}
+		} else if len(gvk.Kind) == 0 {
+			return &NoKindError{fmt.Sprintf("resource #%v has no kind", i)}
+		} else if len(un.GetName()) == 0 {
+			return &NoNameError{fmt.Sprintf("resource #%v has no name", i)}
+		}
+
+		if err := validateResource(un); err != nil {
+			return &ValidationError{fmt.Sprintf("error validating resouce #%v (%v Name=%v Namespace=%v): %v",
+				i, un.GetObjectKind().GroupVersionKind(), un.GetName(), un.GetNamespace(), err)}
+		}
+		success := fmt.Sprintf("resource #%v is valid (%v Name=%v Namespace=%v)\n",
+			i, un.GetObjectKind().GroupVersionKind(), un.GetName(), un.GetNamespace())
+		writer.Write([]byte(success))
```

Comment:
This seems like a somewhat generic parsing method, except for the logging to the writer. Is there similar code elsewhere? I wonder if we should just convert this into a generic parsing method that returns an error or the parsed objects.

Reply:
Looks like we can leverage https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/cli-runtime for this purpose. That will also gives us JSON support, loading remote files (via http(s)), recursive directory loading, etc. 

That requires updating the k8s dependencies (v1.12).

Reply:
Would that also require that we upgrade to the k8s 1.12 runtime? Or could these tools be used independent of k8s version?

Reply:
They're somewhat independent. [k8s version skew policy](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/release/versioning.md#supported-releases-and-component-skew) is +-1 minor version skew for clients against the master. We'll most likely require k8s v1.10 for Istio 1.1.x given (a) k8s v1.12 release was recently announced and (b) Istio supporting the three most recent k8s version. Using k8s 1.11 runtime (deps) allows us to support the min/max supported k8 versions.